text/0005-is-operator.md

```diff
@@ -0,0 +1,176 @@
+Otherwise, the validator gives `e is et` type `Bool`.
+
+To improve typing precision, we can extend the current notion of "effects" to track information about entity types.
+Currently, the validator uses effects to know whether an expression is guaranteed to have an attribute.
+In particular, the validator knows that if `e has a` evaluates to true, then expression `e` is guaranteed to have attribute `a`.
+We can do something similar for `is`: when the expression `e is et` evaluates to true, expression `e` is guaranteed to be an entity of type `et`.
+
+## Other changes
+
+The Dafny formalization of the Cedar language, corresponding proofs, and differential testing framework in [cedar-spec](https://github.com/cedar-policy/cedar-spec) will need to be updated with the new `is` operator.
+
+# Drawbacks
+
+1. This is a substantial new feature that will require significant development effort.
+
+2. This change will break existing applications that depend on the structure of policies. For example, if we decide to allow `is` in the policy scope (as shown in the examples above), tools that perform basic "policy slicing" based on the scope will need to decide how to handle the new syntax.
+
+3. The `is` operator makes individual policies easier to understand because the type information is in the policy itself. However, it may make it more difficult to get a global view of how different actions and entity types are used. The schema currently fills this role: it provides a centralized place to provide all typing constraints.
+
+4. The `is` operator may encourage coding practices that go against our recommended "best practices" (more details below).
+
+## Best practice: Using "typed" actions
+
+From the discussion on <https://github.com/cedar-policy/cedar/issues/94>:
+
+> Guidance we have given other customers is that actions should be "typed," in the sense that they include in their name the type of the resource they apply to. So instead of writing
+>
+>     permit(
+>       principal == User::"admin",
+>       action == Action::"view",
+>       resource is Photo
+>     );
+>
+> you would write
+>
+>     permit(
+>       principal == User::"admin",
+>       action == Action::"viewPhoto",
+>       resource
+>     );
+>
+> In your schema, you indicate that the "viewPhoto" action should always be accompanied by a `Photo` for the resource, so the policy doesn't need to say that explicitly (and doesn't need to perform a runtime check).
+
+If we support an `is` operator, then customers may be encouraged to use a single "view" action for multiple resource types, rather than a new "viewX" actions for each resource type X.
+
+# Alternatives
+
+1. For the most part, users can currently solve the problems that `is` solves by specifying in the schema what types of principals and resources an action applies to, or by using `has` checks, per the examples in the [Motivation](#motivation) above.
+However, in some cases, `is` may result in policies that are easier to read.
```

Comment:
Which cases? Let's see those in the motivation, if the ones there now are just as easily handled with existing mechanisms.

Reply:
This second sentence was referring to "Use case 2"